### PR TITLE
[hackday-remove-dependency] Removing qwest and jparam lib

### PR DESCRIPTION
**Pull request by FernandaBernardo**
This pull request removes dependency on qwest and its dependencies, making ajax-amd deal directly with `ActiveXObject` and `XMLHttpRequest`. Supports get and post requests. Post requests can send data only in `application/x-www-form-urlencoded; charset=UTF-8` format for now.

### CODE REVIEW

File: ajax.js [hunk 1]
```diff
@@ -25,51 +30,103 @@ define('ajax', ['qwest'], function(qwest) {
...
+			return;
+		}
+
+		var responseType = requestObj.getResponseHeader("Content-Type") || "";
+		var responseContent = "";
+
+		if (responseType.indexOf("json") !== -1) {
+			responseContent = JSON.parse(requestObj.responseText);
+		} else if (responseType.indexOf("xml") !== -1) {
+			responseContent = (new DOMParser()).parseFromString(requestObj.responseText,'text/xml');
```
Comment (mottam) [inline ajax.js]: Let's change all `'` to `"`.

@erikatakahara @fellyph

File: ajax.js [hunk 2]
```diff
@@ -25,51 +30,103 @@ define('ajax', ['qwest'], function(qwest) {
...
+		var crossOrigin = host && (host[1] ? host[1] != window.location.host : false);
+
+		headerValues["Accept"] = configHeaders["Accept"] || "*/*";
+
+		if (!crossOrigin) {
+			headerValues['X-Requested-With'] = 'XMLHttpRequest';
+		}
+
+		if (method === POST) {
+			headerValues["Content-Type"] = "application/x-www-form-urlencoded; charset=UTF-8";
```
Comment (mottam) [inline ajax.js]: I think this value should be a default config of this header, but I should be able to overwrite this header for another.

@erikatakahara @fellyph

File: ajax.js [hunk 3]
```diff
@@ -25,51 +30,103 @@ define('ajax', ['qwest'], function(qwest) {
...
+			callbacks.complete(requestObj);
+		}
+	}
+
+	function setHeaders(url, requestObj, method, configHeaders) {
+		var headerValues = {};
+		var host = url.match(/\/\/(.+?)\//);
+		var crossOrigin = host && (host[1] ? host[1] != window.location.host : false);
+
+		headerValues["Accept"] = configHeaders["Accept"] || "*/*";
```
Comment (mottam) [inline ajax.js]: I have this `configHeaders` to overwrite or add new headers to my request, but in this case I can only overwrite the "Accept" config, is this right?

@erikatakahara @fellyph

File: ajax.js [hunk 4]
```diff
@@ -25,51 +30,103 @@ define('ajax', ['qwest'], function(qwest) {
...
+
+		if (responseType.indexOf("json") !== -1) {
+			responseContent = JSON.parse(requestObj.responseText);
+		} else if (responseType.indexOf("xml") !== -1) {
+			responseContent = (new DOMParser()).parseFromString(requestObj.responseText,'text/xml');
+		} else {
+			responseContent = requestObj.responseText;
+		}
+
+		if (isSuccess(requestObj.status) && callbacks.success && typeof callbacks.success === "function") {
```
Comment (erikatakahara) [inline ajax.js]: How about to create `isFunction` that test `callbacks.ANY && typeof callbacks.ANY === "function"`

@tisvasconcelos

**Comment by erikatakahara** [comment]
Today I can pass configs like `retries` and `timeout`, but a didn't see the new implementation using this configs.

@tisvasconcelos 

## 

Done

**Comment by erikatakahara** [comment]
How about to create a test to check if the async config is working? (I don't know if it's possible)

@tisvasconcelos 

## 

done

**Comment by mottam** [comment]
Reminder:
If the response returns empty and the Content-Type is not defined as "application/json", in qwest it's treated as an error when it tries to parse the empty content as Json. On the new lib it behaves correctly and doesn't throw an error,  but we have to remind to change the places where this kind of error is expected.

@erikatakahara

File: ajax.js [hunk 5]
```diff
@@ -25,51 +30,103 @@ define('ajax', ['qwest'], function(qwest) {
...
+
+	function handleResponse(requestObj, callbacks) {
+		if (requestObj.readyState !== 4) {
+			return;
+		}
+
+		var responseType = requestObj.getResponseHeader("Content-Type") || "";
+		var responseContent = "";
+
+		if (responseType.indexOf("json") !== -1) {
```
Comment (mottam) [inline ajax.js]: How about adding try/catch to catch parse exception? In this case if any error occurs at parsing we can call error callback like the others ajax lib.

@erikatakahara 

## 

done

File: ajax.js (hunk 3 above)
Comment (FernandaBernardo) [inline ajax.js]: Done

**Comment by rapahaeru** [comment]
This lib should support CORS as well.

@mottam @luiz 

## 

done

**Pull request merged**